### ISSUE TITLE
Docker Container for TeX Live: Ease of Installation and Font Handling

### ISSUE THREAD

**Issue by thomasWeise** [ustctug/ustcthesis#97]
Dear `ustcthesis` Maintainers.

I recently developed a [Docker](https://www.docker.com/) container which contains a TeX Live installation ([@ DockerHub](https://hub.docker.com/r/thomasweise/texlive/), [@ GitHub](https://github.com/thomasWeise/docker-texlive)). I think something like this (or my container itself) would be a useful thing in the context of ustcslides.

## 1. Advantages: All Software Packages Up-to-Data in Single VM Image
Reason: One of the problems that students often have is installing LaTeX and getting LaTeX to run properly, especially with fonts. With Docker, we can solve this problem.

A Docker container is something like a light-weight virtual machine with a pre-defined image that can run under Windows, Linux, and Mac OS. (Under Linux, it runs as normal process, under Windows and Mac OS, it will run as small VM). This image can contain all required software packages, so no installations of software is required (except for Docker) and everything would be in the right version (including LaTeX packages). Docker is easy to install and the VM can be started with a single command, bringing the user to a command promt.

## 2. Obtaining Docker: Very Easy

Docker can be installed following the guidelines below:

* for [Linux](https://docs.docker.com/linux/step_one/), you can run  `curl -fsSL https://get.docker.com/ | sh` on your command line and everything is done automatically (if you have `curl` installed, which is normally the case),
* for [Windows](https://docs.docker.com/windows/step_one/)
* for [Mac OS](https://docs.docker.com/mac/step_one/)

## 3. How My Image Works: Mount Document and Fonts in 1 Command Line, then Compile

My Docker container works as follows: For compiling some document named `myDocument.tex` in folder `/my/path/to/document/` with (my script) `xelatex.sh` and using fonts in folder `/path/to/fonts/`, you would type something like the command below into a normal terminal (Linux), the *Docker Quickstart Terminal* (Mac OS), or the *Docker Toolbox Terminal* (Windows):

    docker run -v /my/path/to/document/:/doc/ -v /path/to/fonts/:/usr/share/fonts/external/ -t -i thomasweise/texlive
    xelatex.sh myDocument
    exit

This should leave the compiled PDF file in folder `/my/path/to/document/`. If you are not using my pre-defined scripts for building, I recommend doing `chmod 777 myDocument.pdf` after the compilation, to ensure that the produced document can be accessed inside your real (host) system's user, and not just from the Docker container.

### 3.1. How to Compile `ustcthesis`

For the `ustcthesis`template, you would probably replace the above code with something like

    docker run -v /my/path/to/document/:/doc/ -v /path/to/fonts/:/usr/share/fonts/external/ -t -i thomasweise/texlive
    make
    chmod 777 main.pdf
    exit

### 3.2. Windows Fonts

Since Windows such as SimHei fonts are copyrighted, I cannot put them into the container. However, a student may put them into a folder `/path/to/fonts/` and provide option `-v /path/to/fonts/:/usr/share/fonts/external/` to my container. When the container starts, this external fonts path will be mounted as `/usr/share/fonts/external/` and `fc-cache`ed automatically. This seems to work fine.

I think it would even work under Windows, i.e., one could mount something like `C:\Windows\Fonts` directly, but I am not sure about the exact path syntax here, since I don't have a Windows machine. If that would work, I think using such a Docker container would be the easiest way to compile USTC theses.

## 4. Suggestion: Either Use My Image or Make Your Own

You can find my container on [DockerHub](https://hub.docker.com/r/thomasweise/texlive/) and on [GitHub](https://github.com/thomasWeise/docker-texlive). Since it is GPL licensed, you may either use it or create an own, similar container. I think you could even create you own container using my container as base image and have the style and class files of `ustcthesis` copied into the `/doc/` folder. This way, a student would not even need to download them from here anymore. The student would just provide her own text and could compile it with the newest version of the class. (The downside would be that the student would probably no longer preserve a copy of the `ustcthesis` styles, just her thesis `text`, i.e., could maybe not compile it anymore in 20 years from now if `ustcthesis` changes...)

Long story short: A link and short info on how to use a Docker container for TeX Live in your README.md could maybe help students who struggle with LaTeX installations.

**Comment by zepinglee** [comment]
Thanks for your work! 
I noticed the arising use of Docker, but recently I don't have much time trying it.
You are welcomed to join the organization so that you will have a write access.
